stablehlo/conversions/tosa/transforms/StablehloQuantLegalizeToTosaRescale.cpp

```diff
@@ -0,0 +1,257 @@
+/* Copyright 2024 OpenXLA Authors. All Rights Reserved.
```

Comment:
As I was reviewing I noticed  some minor refactoring opportunities. I've put together a patch which is basically a couple of  refactoring I tried and thought it would be useful to share as is avoiding comment threads which might be difficult to track. Feel free to take a look and adopt/modify as you find them useful. Happy to chat further.


https://github.com/openxla/stablehlo/commit/2ae7fd6da16f0b019f3967dc485dd4b42c4e85ed 


Overall the refactoring are related  to
1. Using template over macros. 
2. The actual definition of `matchAndReqrite<AddOp>` or  matchAndRewrite<SubtractOp>` are quite small and IMO the split of decl and defn doesn not add much clarity. 
3. [optional] Expose a `populateStablehloLegalizeQuantToTosaRescalePatterns` public method they can be reused
4. Throw error message if the pass fails 

I will next review the meat (how rescale ops are created) of the conversion logic.

Reply:
done